### PR TITLE
AI - Livestock escape fixed, water jumping fixed

### PR DESCRIPTION
**Pull request by SafwatHalaby**

### CODE REVIEW

**Comment by SafwatHalaby** [comment]
With the exception of `bool cMonster::EnsureProperDestination(cChunk & a_Chunk)`, the changes are trivial.

I did my best to not misuse chunks in that function, but can anyone double-check that?

File: src/Mobs/Monster.cpp [hunk 1]
```diff
@@ -252,6 +259,64 @@ void cMonster::MoveToWayPoint(cChunk & a_Chunk)
...
 
 
+bool cMonster::EnsureProperDestination(cChunk & a_Chunk)
+{
+	cChunk * Chunk = a_Chunk.GetNeighborChunk(m_FinalDestination.x, m_FinalDestination.z);
+	BLOCKTYPE BlockType;
+	NIBBLETYPE BlockMeta;
+	int RelX = m_FinalDestination.x - Chunk->GetPosX() * cChunkDef::Width;
+	int RelZ = m_FinalDestination.z - Chunk->GetPosZ() * cChunkDef::Width;
+	if ((Chunk != nullptr) && Chunk->IsValid())
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Use an early bail-out:
```cpp
if (!cond)
{
  return false;
}

// Do whatever the original if block did
```

File: src/Mobs/Monster.cpp [hunk 2]
```diff
@@ -252,6 +259,64 @@ void cMonster::MoveToWayPoint(cChunk & a_Chunk)
...
+	{
+		// If destination in the air, go down to the lowest air block.
+		for (;;)
+		{
+			Chunk->GetBlockTypeMeta(RelX, m_FinalDestination.y - 1, RelZ, BlockType, BlockMeta);
+			if (cBlockInfo::IsSolid(BlockType))
+			{
+				break;
+			}
+			m_FinalDestination += Vector3d(0, -1, 0);
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: I think it'd be better to use `m_FinalDestination.y += 1` instead, as that will not use a needless new object.

File: src/Mobs/Monster.cpp [hunk 3]
```diff
@@ -252,6 +259,64 @@ void cMonster::MoveToWayPoint(cChunk & a_Chunk)
...
+			else if (cBlockInfo::IsSolid(BlockType))
+			{
+				InWater = false;
+			}
+			else
+			{
+				break;
+			}
+			m_FinalDestination += Vector3d(0, 1, 0);
+		}
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Both loops will fail in a world where the void below is accessible - if there is no solid block under the mob. Use `while (m_FinalDestination.y < cChunkDef::Height)`

File: src/Mobs/Monster.h [hunk 4]
```diff
@@ -204,6 +204,7 @@ class cMonster :
 	*/
 	bool TickPathFinding(cChunk & a_Chunk);
 	void MoveToWayPoint(cChunk & a_Chunk);
+	bool EnsureProperDestination(cChunk & a_Chunk);
```
Comment (madmaxoft) [inline src/Mobs/Monster.h]: Add doxy-comments about what the functions do.

File: src/Mobs/Monster.cpp (hunk 2 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Good point, I have lots of those all over the place, I'll submit a change to all of them in a different PR.

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

File: src/Mobs/Monster.cpp (hunk 3 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

File: src/Mobs/Monster.h (hunk 4 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: Done.

File: src/Mobs/Monster.cpp (hunk 1 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

**Comment by madmaxoft** [comment]
I'd have preferred a doxy-comment without empty lines, but an empty lines between the previous function and the next function's doxy-comment.

**Comment by SafwatHalaby** [comment]
Done.

**Comment by SafwatHalaby** [comment]
Appveyor finished compiling way back but it's stuck. It might have to do with the Github downtime that occured. Nevertheless, all CI's are successful.

**Comment by SafwatHalaby** [comment]
Rebased to shake the CI.

**Comment by SafwatHalaby** [comment]
Things are acting weird, re-opening a new branch.

**Comment by SafwatHalaby** [comment]
#1950

**Pull request closed without merging**